boa/explorer.py

```diff
@@ -1,18 +1,38 @@
 import json
+import logging
+from time import sleep
 from typing import Optional
 
 import requests
 
 SESSION = requests.Session()
 
 
-def _fetch_etherscan(uri: str, api_key: Optional[str] = None, **params) -> dict:
+def _fetch_etherscan(
+    uri: str, api_key: Optional[str] = None, num_retries=10, backoff_ms=400, **params
+) -> dict:
+    """
+    Fetch data from Etherscan API.
+    Offers a simple caching mechanism to avoid redundant queries.
+    Retries if rate limit is reached.
+    :param uri: Etherscan API URI
+    :param api_key: Etherscan API key
+    :params num_retries: Number of retries
+    :params backoff_ms: Backoff in milliseconds
+    :param params: Query parameters
+    :return: JSON response
+    """
     if api_key is not None:
         params["apikey"] = api_key
 
-    res = SESSION.get(uri, params=params)
-    res.raise_for_status()
-    data = res.json()
+    for _ in range(num_retries):
+        res = SESSION.get(uri, params=params)
+        res.raise_for_status()
+        data = res.json()
+        if data.get("result") != "Max rate limit reached":
+            break
+        logging.warning(f"Rate limit reached, retrying in {backoff_ms}ms...")
```

Comment:
i'm not convinced we need logging here

Reply:
I thought it would be useful for the user to know if it takes too long. But yeah 4 seconds is fine.